### ISSUE TITLE
Tasks and Checklist integration

### ISSUE THREAD

**Issue by compgeniuses** [nextcloud/deck#1937]
I is always frustrating that after adding a card to a board, the only way to have itemized content is through comments. or through using the separate task app in nextcloud.
Also, why do we need to have 2 separate apps that do more or less the same functions?
This is confusing even to team members given tasks to do
**What needs to be done**
Tasks as well as checklists needs to be added and integrated into the deck app. this way, when a new card is created and assigned to a user, also a checklist of tasks with items can be added ontop of it, plus other personales can be assigned to the small tasks on that particular card.

**Alternatives**
IF you haev a look at ryver, it has a similar kind of implementation on their kanban boards.

**Additional context**
As seen on this board here. there is acheclist tasks added
![image](https://user-images.githubusercontent.com/16150798/82526232-26d41e00-9b3c-11ea-8ccb-05441d92ba9f.png)

So in this case, what can be done is, the existing task can be integrated as such or you can integrate your own. that way, there is no confusion

**Comment by putt1ck** [comment]
You can add checklists to a card by adding them in the description using - [ ] item

**Comment by compgeniuses** [comment]
That is correct, but it is ot really a task list. which is what i am taalking about.

Between you and i, we can use that process, but for an average office secretary, they d not know markdown, or fancy items.

So embeding the tasks into the board cards on deck, makes alot of sense. and is actually going to be useful.

**Comment by compgeniuses** [comment]
@stefan-niedermann  i see you also encountered a similar problem when it comes to task lsits, anc checkboxes for tasks.
@juliushaertl  Am looking forward to the implementation of the said feature and functionality

**Comment by stefan-niedermann** [comment]
> @stefan-niedermann i see you also encountered a similar problem when it comes to task lsits, anc checkboxes for tasks.

?? what do you mean?

**Comment by compgeniuses** [comment]
> Then one would always have to jump around and the checkboxes wouldn't be visible directly but only after an additional click 😕 .
> 
> Also: wirte permission is write permission.Manipulating check items can be even more dangerous than deleting a card because it is more subtile and people wouldn't see it immediately. I think this is not a question of permissions but only UI / UX therefore.
https://github.com/nextcloud/deck/issues/1953#issuecomment-633184628

As seen above

**Comment by compgeniuses** [comment]
> Writing a checklist is very useful (with this syntax in the description "- [ ]") to manager task list but it is a problem to use because an user need to have modify permission to allow to check the box.
> It is not correct because the user could also delete or modify single task.
> It will be better if we could have a dedicated tab only for check list
https://github.com/nextcloud/deck/issues/1953#issue-623801246

**Comment by stefan-niedermann** [comment]
Well yeah, but i prefer to keep the checklist within the description, because one can have many checklists like

Prerequisites

Some descriptive text.

- [ ] Item 1
- [ ] Item 2

Actual work

This one can will be started as soon as the prerequisites are fulfilled. Where should this text be if we had only a separate task list? We would then need another new UI element to separate stuff.

- [ ] Item 1
- [ ] Item 2

Definition of Done

- [ ] Item 1
- [ ] ...

After all we have a consense to use markdown here and i think we can assume that the users who want to use it, will learn it. Only alternative for me would be a "create checkbox list"-button in the editor toolbar.

**Comment by compgeniuses** [comment]
in as much the existing markdown version remains the same. thats not problem.

What i am proposing is to implement a tab with a separate, dedicated checklist. As shown on the image above how it can look like.

When it comes to developing tools to be used by the masses, never make assumptions. s said, not eveyone using nextcloud is technical, and i recently deployed nc to a law firm, i had to start training them on things they should already know, but its the world.
If there is a way to make it simpler, go ahead and do it,

The other way i mentioned was, to include the existing nextcloud task in deck. that way it makes alot more sense. and there is no need to install deck and task

Please consider

**Comment by grizzee** [comment]
I'm also interested to have a checklist in the cards. Its good to structure the task, project, whatever a little bit.
I used this function in trello very often.

**Comment by stefan-niedermann** [comment]
@grizzee have you read https://github.com/nextcloud/deck/issues/1937#issuecomment-631894788 and can you please be more preceise what exactly is missing from your point of view?

I think it is very unclear for me what the point of the issue is at all:

- Are you just missing a button to simply adding new a check list item
- Do you need a separate area for the check items? (What is the benefit compared of being able to create multiple check lists in the description if there was a nice button and how would you be able to group check items to separate lists?)
- Do you eant task from the other app to be displayed (meaning an integration of the other app)

**Comment by GuillaumeAmat** [comment]
From my point of view, creating checklists in Markdown as in https://github.com/nextcloud/deck/issues/1937#issuecomment-631894788 is the stone age of UX...

> Well yeah, but i prefer to keep the checklist within the description, because one can have many checklists like [...]
> After all we have a consense to use markdown here and i think we can assume that the users who want to use it, will learn it. Only alternative for me would be a "create checkbox list"-button in the editor toolbar.

Wow, I seriously can't believe it. You don't even try to discuss the user needs, as "you prefer" another way to handle checklists.

For what it's worth, I'm glad @compgeniuses created that issue. I found it because I was looking for that feature to be able to leave Trello for good (the only service my 10 years old (own)Nextcloud instance is not handling right now).

If the (non-)discussion continues that way, I'm not even close to look after that module again.

Good luck @compgeniuses ✊

cc @juliushaertl

**Comment by grizzee** [comment]
@stefan-niedermann i read the comment. But i think the same about this topic like @compgeniuses. To write Checklist item with `- [ ]` is a long and not very comfortable way in my opinion. To have a Tab with a Checklist would be very nice way to work with Cards in Deck.
- a button for a checklist item would good at first for me.
- or a Checklist Area would be even better
- an integration in another is a little bit to much and could be confusing.

**Comment by stefan-niedermann** [comment]
> Wow, I seriously can't believe it. You don't even try to discuss the user needs, as "you prefer" another way to handle checklists.

Hey, sure i am trying to discuss, that's why i asked the questions 🙂 . I am not even a developer of this repository. Therefore i wanted to share my personal preference - just like you. I *am* a user of the web app, too.

> If the (non-)discussion continues that way, I'm not even close to look after that module again.

Well then i better be silent, it wasn't my intention to offend anybody. I really just wanted to understand what you would prefer - tabs, buttons to add checkboxes. I just tried to tell you some advantages of my personal opinion.
about the consense i meant the "status quo" - Notes, Deck, even the Files app, they all are using markdown currently. Sorry, i am not a native english speaker.

Thank you @grizzee that's just what i asked :)

**Comment by putt1ck** [comment]
> From my point of view, creating checklists in Markdown as in [#1937 (comment)](https://github.com/nextcloud/deck/issues/1937#issuecomment-631894788) is the stone age of UX...

Yet it isn't. All that is lacking for those who haven't learned Markdown (just extended punctuation, like emoticons) is better documentation and a button in the editor that generates the initial item. From then on it's quicker and easier than a button; a button requires you to move your hands from the keyboard, when all you want to do is type the checklist items. Once you've learned the punctuation you just keep typing `- [ ] item 1` and from then on in it's just an auto-list, double-enter to end.

**Comment by compgeniuses** [comment]
Lest us understand one thing here. please.

There is a different between a checklist and a check task list.

The markdown checklist implement is ok, but for the average joe, even for me, it does not cut it. I am technical enough, with over 7+ years in system architecture.

It will be wise to implement a separate tab on the cards, which handles checklists task.

Have you seen the image above, that kind of implementation is grate, and anyone can be able to use it effectively.

Plus, implementing that separately, means one does not need to edit the content just to be able to modify the task lists.

That is why i said, it if seems to be a pain to build such a feature, why nto engage the NC task team, to implement the task feature inside the deck. that way, each card on a deck, has a task list, thats beautifully done.

**Comment by putt1ck** [comment]
The trouble is you're asking for it to be made harder for those who've learnt a simple start point, of typing

- [ ] an item

(note once you've created the checklist the items can be checked just by clicking the checkbox)

any change to UI that makes something harder for those who know how to use a tool is poor UX. Users have to learn many things to use any software (and hardware), whether it's "press this button to make text be in bold" or "type - [ ] " to start a checklist.

It could be made easier to learn by having a "start check list" button, just like Github comments do.

**Comment by GuillaumeAmat** [comment]
@stefan-niedermann As you are the main developer of the Deck Android app, your opinion has a lot more weight. Every web feature is inevitably linked to a potential mobile feature. Thank you for the clarification of your point anyway.

> any change to UI that makes something harder for those who know how to use a tool is poor UX

@putt1ck I don't understand why it would be harder to use the Markdown way of handling checklists... That issue is about *adding* a feature, not removing the old one.

I personaly love Markdown and use it heavily every day on several supports, but it is still a "mid-power user" thing (if I may), even with an "add" button. And it can't evolve (eg: link an item to a resource, timestamp, anything meta).

**Comment by putt1ck** [comment]
@GuillaumeAmat the current checklist support is Markdown, same format as a Github comment `- [ ]` which renders as
- [ ] an item
(in earlier versions of Deck the Markdown library in use allowed just "[ ]" and the change from that was confusing, for about 5 minutes)

**Comment by GuillaumeAmat** [comment]
> @GuillaumeAmat the current checklist support is Markdown, same format as a Github comment - [ ] which renders as [...]

What's your point? Did you read my comment?

**Comment by compgeniuses** [comment]
Github is for developers, people who either know how to code, or know about coding, lets not use github as reference. i would rather use trello as the reference, because secretaries and other people use it for their day to day life.

Do not remove the current markdown implementation of checklist, leave it as it is. power users like myself and others will still make use of it.

Just implement a simpler version of tasks. which as suggested, instead of building a new one, you can use the already existing nextcloud tasks, if there is a possibility

**Comment by joreiff** [comment]
Just curious: Are you missing any features in the current markdown-based implementation, e.g. assigning individual tasks of the same card to different people? Or is this purely about UI/UX to make Deck easier to use for non-technical people?

**Comment by compgeniuses** [comment]
1, the checklist is not separated form the main content. needs ability to separate this.
2 someone with no edit function on the list, can not create checklist.
3. if i need someone to be able to add checklist but ability to edit the content of the text, how can i do that.
4. does it have the ability to assign specific task list to a specific person
@joreiff  there are so many different things that the standard checklist can not, and does not handle well.
Which is why i posted the suggestion and recommendation.

But if technical members and developers feel, its good the way it is, then so be it.
All i say is, there is no need to be rigid in what you hold dear, and not understand the actual problems that need to be tackled and added to improve the system.

**Comment by juliushaertl** [comment]
I fully agree on the usability issue that markdown has itself. However there is work being done on integrating Nextcloud Text into Deck, so that users will have WYSIWYG editor with support for checkboxes so adding a todo item is just clicking a button. However my main concern is the huge additional code that such a separation would bring into Deck that we need to maintain.

> 1. the checklist is not separated form the main content. needs ability to separate this.

This is really just a personal preference you have.

> 2. someone with no edit function on the list, can not create checklist.
> 3. if i need someone to be able to add checklist but ability to edit the content of the text, how can i do that.

I'd say that is expected and would be the same case for a separate checklist. Permission management is a totally different topic. Until now we really tried to keep permissions as simple as possible and I don't like the idea of adding another level of complexity with different types of editing permissions. This will make it harder to understand for users as well as a lot more complex to test and maintain.

> 4. does it have the ability to assign specific task list to a specific person

Not yet, but of course that is something that could also be worked on as enhancements to text https://github.com/nextcloud/text/issues/92

From the above discussion I don't really see any point feature that could not be implemented in a visual editor.

> But if technical members and developers feel, its good the way it is, then so be it.
> All i say is, there is no need to be rigid in what you hold dear, and not understand the actual problems that need to be tackled and added to improve the system.

This is not about blocking ideas but understanding the actual requirements and issues of the current solution, that are different from a concrete implementation suggestion as just replicating the trello behavior would be.

**Comment by compgeniuses** [comment]
@juliushaertl Then, as you have put it. I think i have a suggestion on how this can be done.
1) On the projects, implement tasks as one of the projects, so that it is possible to link a given deck board to a specific nextcloud task.

2) For those who do not use task, here is another solution.
Using the same checklist of the markdown. implement a tab or something similar, that displays the checklist outside the Main content.

**Comment by grossherr** [comment]
For what it's worth, but I think integrating tasks with deck is an important feature. IMHO it shouldn't only link to tasks though. The ability to see them at a card and e.g. mark them as done would be pretty great.

**Comment by compgeniuses** [comment]
@grossherr That is exactly what i was requesting.

The ability to see nextcloud Tasks into Deck, and the ability to view them inside a Deck is important.

Also, the ability to create the Said tasks inside the Deck. not necessarily having to go into tasks to create them. that would be a superb plus.
@juliushaertl  @joreiff @stefan-niedermann @putt1ck

**Comment by duesseldorferjung** [comment]
Good Morning Everyone,
I have been a proud user of Deck since the beta versions both Webapp and Android. The progress made is amazing! 
The feature of being able to enter checklist items will be quite helpful for a brooder audience. My observation is, that the agile / kanbo / scrum or however you want to call it, way of working and thinking is now implemented much beyond IT Nerds. My believe is that Opensource should be there for everyone and therefore should give a good user experience too everyone.

Typical usecase for me is, I will have a kanbocard per Shop I am visiting on different days to buy groceries. Within the description I just want to be able to really quickly enter everything I need and be able to check it off. (Hitting enter should create a new Checklist item in the next line). Having a button to switch between normal and checklist mode would be an option for me.

The same feature - in Microsoft Kanbo - is actually the most used one in my IT Team. (So far nobody has complain that they can not add it in markup language).

**Comment by putt1ck** [comment]
@duesseldorferjung checklist items are already available by typing `- [ ] item`

**Comment by duesseldorferjung** [comment]
Thanks @putt1ck I am aware of that and I share the opinion of others in this thread that giving a different option would make it more user friendly.

**Comment by putt1ck** [comment]
A use not knowing how to do something doesn't mean the application is user unfriendly. It just means the user doesn't know how to do something. While a button a la Github comments to start a checklist would be useful, it's far quicker to just type the first `- [ ]` than to take your hands off the keyboard to press a button. It's worth noting that Github and others use the buttons to show the notation i.e. to teach you how to do it.

NB just pressing enter at the end of the first checklist item creates a new checklist item, just like most auto-list implementations in word processors.

**Comment by juliushaertl** [comment]
Let me close this as our current roadmap related to enhance the checklist/task-list behavior is already tracked in the following tickets:

- #1561 will show a button and have rich editing in the card description
- #15 for integrating the deck app with caldav and nextcloud tasks, where it has not been decided yet if and how subtasks will be mapped to deck.
